Add TwoWay bindable property PageIndex

jldearmas opened the pull request:
Added property **PageNumber** that can be bound with Mode=TwoWay; so it allows to keep updated the page number, but also go to an specific page.

The example was modified to show the usage of the **PageNumber** property using Label and Stepper controls:

```
<Label
    Grid.Row="0"
    Grid.Column="2"
    FontSize="Caption"
    HorizontalTextAlignment="Center"
    Text="{Binding PageNumber, StringFormat='Page number is {0}'}" />

<Stepper
    Grid.Row="1"
    Grid.Column="2"
    Maximum="5"
    Minimum="1"
    Value="{Binding PageNumber, Mode=TwoWay}" />
```

```
<pdf:PdfView
    Grid.Row="1"
    IsHorizontal="{Binding IsHorizontal}"
    MaxZoom="{Binding MaxZoom}"
    PageChangedCommand="{Binding PageChangedCommand}"
    PageNumber="{Binding PageNumber, Mode=TwoWay}"
    Uri="{Binding PdfSource}"/>
```

jldearmas commented:
Testing in Android

![image](https://github.com/user-attachments/assets/13217121-849b-4c5a-bf88-9d8afb48684b)

vitalii-vov commented:
Hi @jldearmas can you answering for the questions about parts of code what I requested?

jldearmas commented:
OK... No problem, except I can't find the questions. Where are them?

vitalii-vov commented:
The code review requests are present on this page, but they may be collapsed in your view, which could explain why they are not visible.

<img width="1048" alt="Screenshot 2025-05-14 at 7 32 24 PM" src="https://github.com/user-attachments/assets/7344644a-02f4-40e4-a135-7a933b1629a9" />

jldearmas commented:
GitHub is working different for me, see my page, there is no code review section

![image](https://github.com/user-attachments/assets/74df9cfb-ec21-489d-a110-aeeb3c0c3ff3)

jldearmas commented:
Anyway, I saw your questions in the screenshot that you post:

1. About **skipping pageNumber 0**. The matter is that PageNumber is not a page index, but the printed page number, so it start at 1 (not at 0)
2. About **setting PageNumber when zoomed in** for Android, Mac and iOS there is no problem. In the case of Windows Platform there was a bug that I fixed. Doing that I also found and fixed another issue (for all platforms) related with update PageNumber during scroll.

jldearmas commented on Maui.PDFView/Platforms/Android/PdfViewHandler.cs:
```diff
@@ -157,7 +163,17 @@ void RenderPages()
+        {
+            var layoutManager = (LinearLayoutManager)_recycleView.GetLayoutManager()!;
+
+            if (pageNumber == 0 || layoutManager.ItemCount <= pageNumber - 1)
```
The matter is that PageNumber is not a page index, but the printed page number, so it start at 1 (not at 0)

vitalii-vov replied:
I believe we should avoid handling view-related logic (like page number) outside the `View` layer. Instead, we should just work with indexes and let the View decide how to present them.
Otherwise, we end up introducing unnecessary UI-specific operations across the codebase, such as 
```C#
layoutManager.ScrollToPosition((int)pageNumber - 1);
```

vitalii-vov replied:
It would be great if the line
```C#
if (pageNumber == 0 || layoutManager.ItemCount <= pageNumber - 1)
```
be changed to the string
```C#
if (pageNumber >= layoutManager.ItemCount)
```
in all platform renderers.

jldearmas replied:
**Done!!**

jldearmas commented on Maui.PDFView/Platforms/Android/PdfViewHandler.cs:
```diff
@@ -209,8 +225,19 @@ public override void OnScrolled(RecyclerView recyclerView, int dx, int dy)
 
-                // Execute the command if available
-                _handler.VirtualView.PageChangedCommand?.Execute(new PageChangedEventArgs(currentPage + 1, layoutManager.ItemCount));
+                if (currentPage >= 0)
```
My bad, that's not necessary, I'll remove it.

jldearmas replied:
Done

vitalii-vov replied:
It still available on [Windows](https://github.com/jldearmas/Maui.PDFView/blob/6afa614b0214ebe16e3879bb2386af768dc9924e/Maui.PDFView/Platforms/Windows/PdfViewHandler.cs#L229)

jldearmas replied:
**Done!!**

vitalii-vov commented on Maui.PDFView/Platforms/iOS/PdfViewHandler.cs:
```diff
@@ -126,8 +126,13 @@ private void RenderPages()
         private void GotoPage(uint pageNumber)
         {
+            if (isScrolling)
+                return;
```
What problem does this piece of code solve?
If you need to stop scrolling while selecting a page, maybe it's worth considering stopping the scroll animation with `scrollView.setContentOffset(scrollView.contentOffset, animated: false)` before setting the page index?

jldearmas replied:
The matter is that set the scroller to a page's top (at least in Windows Platform, and seems the same in Android) align the page's top with the top of the scroller, so when the user scroll the PDF, if the GotoPage is executed, there is a weird visual effect; so that code it's to avoid that. It doesn't have to do with scroll animation.

vitalii-vov replied:
As I see it, there may be a situation of missing a command call, in case of calling it sequentially from the code. For example, if at the moment of `GotoPage` call `isScrolling == true`.

vitalii-vov commented on Maui.PDFView/Platforms/Windows/PdfViewHandler.cs:
```diff
@@ -171,8 +176,7 @@ private void GotoPage(uint pageNumber)
 
-                //this.pageNumber = pageNumber;
-                _scrollViewer.ChangeView(child.ActualOffset.X, child.ActualOffset.Y, 1, true);
+                _scrollViewer.ScrollToVerticalOffset(_scrollViewer.ZoomFactor*child.ActualOffset.Y);
```
Does this work for horizontal scrolling as well?

jldearmas replied:
I fixed in a version I'll push soon

jldearmas replied:
Done

vitalii-vov commented on Maui.PDFView/Platforms/Windows/PdfViewHandler.cs:
```diff
@@ -191,9 +216,17 @@ private void OnScrollViewerViewChanged(object sender, ScrollViewerViewChangedEve
+                    if (VirtualView.PageChangedCommand?.CanExecute(null) == true)
+                    {
+                        VirtualView.PageChangedCommand.Execute(new PageChangedEventArgs(currentPage + 1, layout.Children.Count));
+                    }
```
As I see from the code, `currentPage + 1` was originally added by me so that `PageChangedEventArgs` would display the page number instead of the index, so this mistake was initially mine, and you probably followed that example. I think in the future I’ll make a change to fix this and replace `CurrentPage` with `CurrentPageIndex` or something like that.

jldearmas commented on Maui.PDFView/Platforms/Android/PdfViewHandler.cs:
```diff
@@ -217,8 +233,16 @@ public override void OnScrolled(RecyclerView recyclerView, int dx, int dy)
+                    if (_handler.VirtualView.PageNumber != newPage)
+                        _handler.VirtualView.PageNumber = newPage;
+
+                    if (_handler.VirtualView.PageChangedCommand?.CanExecute(null) == true)
```
That code above (using CommunityToolKit .MVVM), create a **Command** that implemented both methods:
- CanExecute
- Execute
with the actions/delegates supplied. If the programmer don't supply an action for implement **CanExecute**, this method returns true.

But the code that that call a **Command** has the responsibility of check with **CanExecute**.

jldearmas commented:
**Windows Platform**

![image](https://github.com/user-attachments/assets/2866caa5-2329-48c3-8d47-59a92125b6fb)

jldearmas commented:
**Android Platform**

![image](https://github.com/user-attachments/assets/3c989fdf-7c43-4c9c-992b-dfbfb15e1dde)

vitalii-vov commented:
Hello @jldearmas ,
I'm trying to test a new feature and encountered some strange behavior on Android, which I demonstrated in the video.

As shown in the video, when switching pages, we sometimes don't receive a callback from the control.
You can see that the page numbers change like 1/5, 3/5, 4/5, etc., meaning that 2/5 is being skipped.
The OnScrolled function is not always triggered.

[android-page-index-bug.webm](https://github.com/user-attachments/assets/64fdcb27-70a7-4c74-a24a-5580dcd89fd9)

vitalii-vov commented on Maui.PDFView/Platforms/Android/PdfViewHandler.cs:
```diff
@@ -165,7 +171,24 @@ void RenderPages()
-        
+
+        private bool isScrolling;
+        private bool isPageIndexLocked;
```
Private variables must be declared according to the general coding style, that is, starting with an underscore:
```C#
private bool _isScrolling;
```
instead of 
```C#
private bool isScrolling;
```

They should also be declared before the constructor to maintain the overall code structure.
This rule applies to all code, not just this part.

vitalii-vov replied:
Fixed

jldearmas commented:
So, is there something pending?

vitalii-vov commented:
Overall, this build looks ready for release on `iOS`, `macOS`, and `Android`, but I haven’t tested it on `Windows` yet. I didn’t have enough time for that today, but I’ll check it as soon as possible.

Also, I’ve been thinking that the `MaxPageIndex` property you initialize [here](https://github.com/jldearmas/Maui.PDFView/blob/58d1dc24c96ded1dcf4375530fbb5b8eeecb955d/Example/Business/UI/ViewModels/MainPageViewModel.cs#L33) could potentially be a `TwoWay` bindable property of the `PdfView` component. However, I’m not entirely sure whether it’s worth modifying the library for that, or if it’s better to accept it as is. 

But in general, I think after testing on Windows, I’ll accept your contribution, as it could be quite useful for many users.

vitalii-vov commented:
There is some problem on Windows — when we change Uri Current index not change

jldearmas commented:
Thanks, was a pleasure work with you

vitalii-vov commented:
Thank you, @jldearmas, for your contribution to the library. Your involvement in this feature is greatly appreciated.
